### PR TITLE
Fill in the blanks exercise prototype

### PR DESCRIPTION
**Pull request by LarsTheGlidingSquirrel**
Prototype for fill-in-the-blanks exercise

Preview: [link](https://frontend-git-fill-in-the-gap-exercise-prototype-serlo.vercel.app/___editor_preview?name=%7B%22plugin%22%3A%22rows%22%2C%22state%22%3A%5B%7B%22plugin%22%3A%22exercise%22%2C%22state%22%3A%7B%22content%22%3A%7B%22plugin%22%3A%22rows%22%2C%22state%22%3A%5B%7B%22plugin%22%3A%22text%22%2C%22state%22%3A%5B%7B%22type%22%3A%22p%22%2C%22children%22%3A%5B%7B%22text%22%3A%22F%C3%BClle+die+L%C3%BCcken+aus%3A+%22%7D%5D%7D%5D%2C%22id%22%3A%22c6977032-9ea7-4ecd-ab92-4bf4451af892%22%7D%5D%2C%22id%22%3A%22b663090d-533a-4f09-8a72-911e53996b5a%22%7D%2C%22interactive%22%3A%7B%22plugin%22%3A%22fillInTheGapExercise%22%2C%22state%22%3A%7B%22text%22%3A%7B%22plugin%22%3A%22text%22%2C%22state%22%3A%5B%7B%22type%22%3A%22p%22%2C%22children%22%3A%5B%7B%22text%22%3A%22Ich+bin+ein+%22%7D%2C%7B%22type%22%3A%22gap%22%2C%22gapId%22%3A%222c7fe147-9d8c-4839-9422-68da67926a7c%22%2C%22correctAnswer%22%3A%22Text%22%2C%22alternativeSolutions%22%3A%5B%5D%2C%22children%22%3A%5B%7B%22text%22%3A%22%22%7D%5D%7D%2C%7B%22text%22%3A%22+mit+L%C3%BCcken%22%7D%5D%7D%5D%2C%22id%22%3A%2261f9758b-04a6-48b5-a9b0-146aca861bca%22%7D%2C%22mode%22%3A%22fill-in-the-gap%22%7D%2C%22id%22%3A%22e92fe9f8-624b-4db6-807d-f519ed4d44f0%22%7D%7D%2C%22id%22%3A%22e46938da-0b10-4fcd-a2e7-4ae121d988c7%22%7D%2C%7B%22plugin%22%3A%22solution%22%2C%22state%22%3A%7B%22strategy%22%3A%7B%22plugin%22%3A%22text%22%2C%22state%22%3A%5B%7B%22type%22%3A%22p%22%2C%22children%22%3A%5B%7B%22text%22%3A%22%22%7D%5D%7D%5D%2C%22id%22%3A%22fca81afa-3465-4751-8550-681b5ef5dae3%22%7D%2C%22steps%22%3A%7B%22plugin%22%3A%22rows%22%2C%22state%22%3A%5B%7B%22plugin%22%3A%22text%22%2C%22state%22%3A%5B%7B%22type%22%3A%22p%22%2C%22children%22%3A%5B%7B%22text%22%3A%22L%C3%B6sung%22%7D%5D%7D%5D%2C%22id%22%3A%22052a137f-3a56-4a9f-b28a-c5a74639672f%22%7D%5D%2C%22id%22%3A%228c14c7e1-6140-4b95-a7b8-950a2c02b519%22%7D%7D%2C%22id%22%3A%2290f2bcf8-7ad3-4948-8c4d-a5c346d165b2%22%7D%5D%7D)

Not ready for production yet -> Should only be available in staging for now

Important to know before reviewing:
- Exercise has two modes: 
(1) Fill-in-the-blanks mode -> User types into blanks. Fully works on a basic level. 
(2) drag-and-drop mode -> User drags answers into blanks. Does not work fully. I only experimented a bit if I can get it working with the current implementation. You might see commented out code for that mode. Just ignore for now. 
- My diagram about the implementation might help understand how it is implemented: https://miro.com/app/board/uXjVMgBV4FM=/?moveToWidget=3458764568180068650&cot=14
- Let me know if it would be more helpful to have a meeting where I give an overview. :) 

Specific things to review:
- Is rendering split up correctly into static and editor rendering? ping @elbotho 
- Is hiding in production done right? I simply check `isProduction` and remove the fill-in-the-blanks exercise plugin in `create-plugins.ts`, `create-renderers.rs` and `src/serlo-editor/plugins/exercise/editor.tsx`

TODOs - before merging
- [x] Make blanks with no content smaller
- [x] Investigate bundle size increase and try to make it smaller
- [x] Use correct name instead of "fill-in-the-gap exercise" -> Now "Fill-in-the-blanks"
- [x] Ignore white spaces (start/end) when checking solution

TODOs - probably after this PR is merged but we can also do them before if it makes sense:
- [ ] Only display "Stimmts?" if all blanks have content? 
- [ ] Make clicking "Stimmts?" in editor working. (Probably fixed by using BlankRenderer here as well)
- [ ] Use react-dnd instead of dnd-kit to have one less dependency?
- [ ] Find better method to size input elements depending on content
- [ ] Moving in and out of blanks using arrow keys.
- [ ] Drag & drop mode:
  - [ ] Generate draggables from text state and make that work
  - [ ] Check correct answer
- [ ] Toolbar for exercise. Nested like in multimedia.
- [x] Make blank button in toolbar look better
- [ ] Removing blank should keep text instead of deleting it.
- [ ] Remove entries in textUserTypedIntoBlanks when blanks get deleted in text state
- [ ] Hide formatting options (bold ...) when selection contains blank
- [ ] Make TextEditor also use BlankRenderer
- [ ] Make Text plugin toolbar display "Lückentext" instead of "Text"
- [ ] Change visuals according to design

### CODE REVIEW

**Comment by elbotho** [comment]
Some testing stuff, if that's helpful:

The default gap size seems a big big for me
image

We could make it move obvious what the learner has to do by not showing the check button before all gaps are filled and instead show a task (like we do for single choice exercises):
image

Side margin of check button too small (might also be because of the preview, but works for other exercises)
image

File: src/data/en/index.ts [hunk 1]
```diff
@@ -890,6 +891,10 @@ export const loggedInData = {
         scMcExercise: {
           title: 'SC/MC Exercise',
           description: 'Single Choice or Multiple Choice'
+        },
+        fillInTheGap: {
+          title: 'Fill-In-The-Gap Exercise',
```
Comment (elbotho) [inline src/data/en/index.ts]: I just checked and actually that's not correct english :)
It probably would be `cloze` or maybe `fill in the blanks`.

File: src/serlo-editor/plugins/text/hooks/use-text-config.tsx [hunk 2]
```diff
@@ -10,6 +10,7 @@ const defaultFormattingOptions: TextEditorFormattingOption[] = [
   TextEditorFormattingOption.math,
   TextEditorFormattingOption.richTextBold,
   TextEditorFormattingOption.richTextItalic,
+  TextEditorFormattingOption.gap,
```
Comment (elbotho) [inline src/serlo-editor/plugins/text/hooks/use-text-config.tsx]: Should this actually be in the `defaultFormattingOptions`?

File: src/serlo-editor/plugins/fill-in-the-gap-exercise/index.tsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+          TextEditorFormattingOption.colors,
+          TextEditorFormattingOption.headings,
+          TextEditorFormattingOption.katex,
+          TextEditorFormattingOption.links,
+          TextEditorFormattingOption.lists,
+          TextEditorFormattingOption.math,
+          TextEditorFormattingOption.paragraphs,
+          TextEditorFormattingOption.richTextBold,
+          TextEditorFormattingOption.richTextItalic,
+          TextEditorFormattingOption.gap,
```
Comment (elbotho) [inline src/serlo-editor/plugins/fill-in-the-gap-exercise/index.tsx]: could we extend this from the `defaultFormattingOptions`?

File: src/serlo-editor/editor-ui/plugin-toolbar/text-controls/utils/math.ts [hunk 4]
```diff
@@ -42,7 +42,7 @@ export function toggleMath(editor: SlateEditor) {
         type: 'math',
         src: SlateEditor.string(editor, trimmedSelection as Location) || '',
         inline: true,
-        children: [],
+        children: [{ text: '' }],
```
Comment (elbotho) [inline src/serlo-editor/editor-ui/plugin-toolbar/text-controls/utils/math.ts:45]: why is this change needed?

File: src/serlo-editor-integration/create-plugins.tsx [hunk 5]
```diff
@@ -71,7 +73,21 @@ export function createPlugins({
...
+          TextEditorFormattingOption.code,
+          TextEditorFormattingOption.colors,
+          TextEditorFormattingOption.headings,
+          TextEditorFormattingOption.katex,
+          TextEditorFormattingOption.links,
+          TextEditorFormattingOption.lists,
+          TextEditorFormattingOption.math,
+          TextEditorFormattingOption.paragraphs,
+          TextEditorFormattingOption.richTextBold,
+          TextEditorFormattingOption.richTextItalic,
```
Comment (elbotho) [inline src/serlo-editor-integration/create-plugins.tsx]: can we use could the `defaultFormattingOptions` here?

File: src/data/en/index.ts (hunk 1 above)
Comment (elbotho) [reply src/data/en/index.ts]: This would obviously change all the comp and file names too 😬 but it's probably our last chance to do so :)

**Review by elbotho — CHANGES_REQUESTED** [review]
Thanks @LarsTheGlidingSquirrel !
First round, did not go through everything yet.

**Comment by elbotho** [comment]
Also the  `(🟡 +5.53 KB)` increase in the static renderer is a bit unexpected. Let's see if we can decrease this.

File: src/data/en/index.ts (hunk 1 above)
Comment (LarsTheGlidingSquirrel) [reply src/data/en/index.ts]: Oh! To be honest, I never checked and just went with the name we were using at serlo. I'll check that and yes, we should change it now instead of later. :smile:

File: src/serlo-editor/plugins/text/hooks/use-text-config.tsx (hunk 2 above)
Comment (LarsTheGlidingSquirrel) [reply src/serlo-editor/plugins/text/hooks/use-text-config.tsx]: No, that should not stay in the defaults. Probably there, because it was easier to do it that way initially.

File: src/serlo-editor/plugins/fill-in-the-gap-exercise/index.tsx (hunk 3 above)
Comment (LarsTheGlidingSquirrel) [reply src/serlo-editor/plugins/fill-in-the-gap-exercise/index.tsx]: Good idea!

File: src/serlo-editor/editor-ui/plugin-toolbar/text-controls/utils/math.ts (hunk 4 above)
Comment (LarsTheGlidingSquirrel) [reply src/serlo-editor/editor-ui/plugin-toolbar/text-controls/utils/math.ts:45]: As far as I understand, slate will place an empty text element there when running normalizations. It isn't need at all in this pull request. I should have left a comment here. We can remove it if you want.

Comment (elbotho) [reply src/serlo-editor/editor-ui/plugin-toolbar/text-controls/utils/math.ts:45]: so you are saying that slate will include this anyway if we don't provide it?

Comment (LarsTheGlidingSquirrel) [reply src/serlo-editor/editor-ui/plugin-toolbar/text-controls/utils/math.ts:45]: Yes! It is also one of slates "rules", see https://docs.slatejs.org/concepts/11-normalizing (first rule)

File: src/serlo-editor/plugins/fill-in-the-gap-exercise/components/gap-solution-area.tsx [hunk 6]
```diff
@@ -0,0 +1,17 @@
+// import { useDroppable } from '@dnd-kit/core'
```
Comment (hejtful) [inline src/serlo-editor/plugins/fill-in-the-gap-exercise/components/gap-solution-area.tsx]: Leftover?

File: src/serlo-editor/plugins/fill-in-the-gap-exercise/components/gap-solution.tsx [hunk 7]
```diff
@@ -0,0 +1,27 @@
+// import { UniqueIdentifier, useDraggable } from '@dnd-kit/core'
```
Comment (hejtful) [inline src/serlo-editor/plugins/fill-in-the-gap-exercise/components/gap-solution.tsx]: Leftover?

File: src/serlo-editor/plugins/fill-in-the-gap-exercise/context/gap-context.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import { createContext } from 'react'
+
+type GapId = string
```
Comment (hejtful) [inline src/serlo-editor/plugins/fill-in-the-gap-exercise/context/gap-context.ts]: Could you maybe explain the need for this type declaration? I can't think of a reason, but probably I'm missing something 😄

File: src/serlo-editor/plugins/fill-in-the-gap-exercise/context/gap-context.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+// Used to pass state to GapRenderer from FillInTheGapRenderer
+// GapRenderer will use this state alongside the state stored in the slate custom element 'gap' to render.
+export const GapStatesContext = createContext<{
+  mode: string
+  gapFeedback: Map<GapId, { isCorrect: boolean | undefined }>
```
Comment (hejtful) [inline src/serlo-editor/plugins/fill-in-the-gap-exercise/context/gap-context.ts]: Super minor:
```suggestion
  gapFeedback: Map<GapId, { isCorrect?: boolean }>
```

File: src/serlo-editor/plugins/fill-in-the-gap-exercise/context/gap-context.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+  textUserTypedIntoGap: {
+    value: Map<GapId, { text: string }>
+    set: React.Dispatch<React.SetStateAction<Map<GapId, { text: string }>>>
```
Comment (hejtful) [inline src/serlo-editor/plugins/fill-in-the-gap-exercise/context/gap-context.ts]: Also minor:
```typescript
type UserInput = Map<GapId, { text: string }>
```
```suggestion
    value: UserInput
    set: React.Dispatch<React.SetStateAction<UserInput>>
```

File: src/serlo-editor/plugins/text/hooks/use-slate-render-handlers.tsx [hunk 11]
```diff
@@ -80,6 +80,31 @@ export const useSlateRenderHandlers = ({
...
+        // <GapRenderer> needs to ...
+        // - always show input fields
+        // - accept callback onChange that is passed to input element to update correctAnswer
+        return (
+          <span {...attributes} contentEditable={false}>
+            <input
+              value={element.correctAnswer}
+              size={element.correctAnswer.length}
+              className="rounded-full border border-brand bg-brand-50 px-2"
+              onChange={(evt) => {
```
Comment (hejtful) [inline src/serlo-editor/plugins/text/hooks/use-slate-render-handlers.tsx]: Super minor, but we already have `e` and `event` throughout the codebase, so maybe just use one of those instead of `evt`, for the sake of consistency.

File: src/serlo-editor/plugins/text/hooks/use-slate-render-handlers.tsx [hunk 12]
```diff
@@ -80,6 +80,31 @@ export const useSlateRenderHandlers = ({
...
+      if (element.type === 'gap') {
+        // TODO: Render <GapRenderer> here instead
+        // <GapRenderer> needs to ...
+        // - always show input fields
+        // - accept callback onChange that is passed to input element to update correctAnswer
+        return (
+          <span {...attributes} contentEditable={false}>
+            <input
+              value={element.correctAnswer}
+              size={element.correctAnswer.length}
```
Comment (hejtful) [inline src/serlo-editor/plugins/text/hooks/use-slate-render-handlers.tsx]: What I noticed while using Duolingo, is that they will always add a bit of size, to not make it too obvious what the solution is. So, maybe like `size={element.correctAnswer.length + 2}` or something.

File: src/serlo-editor/plugins/fill-in-the-gap-exercise/renderer.tsx [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+    }
+  }
+
+  // Searches for gap objects in text plugin state. They can be at varying depths.
+  function getGapsWithinObject(obj: object) {
+    if (GapState.is(obj)) {
+      return [obj]
+    }
+
+    // Recursively search this objects values for gap objects
```
Comment (hejtful) [inline src/serlo-editor/plugins/fill-in-the-gap-exercise/renderer.tsx]:
```suggestion
    // Recursively search this object's values for gap objects
```

**Comment by LarsTheGlidingSquirrel** [comment]
> Also the (🟡 +5.53 KB) increase in the static renderer is a bit unexpected. Let's see if we can decrease this.

@elbotho I guess because of importing `io-ts` in the renderer. But not sure. I changed it so that `FillInTheGapRenderer` is imported dynamically. Probably a good idea anyway once the plugin uses some drag & drop library in the next iteration.

File: src/serlo-editor/plugins/text/hooks/use-text-config.tsx (hunk 2 above)
Comment (LarsTheGlidingSquirrel) [reply src/serlo-editor/plugins/text/hooks/use-text-config.tsx]: Simplified the configuration

File: src/serlo-editor-integration/create-plugins.tsx (hunk 5 above)
Comment (LarsTheGlidingSquirrel) [reply src/serlo-editor-integration/create-plugins.tsx]: Simplified the configuration

File: src/serlo-editor/plugins/fill-in-the-gap-exercise/renderer.tsx [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+    let objList: t.TypeOf<typeof GapState>[] = []
+    Object.values(obj).forEach((_value) => {
+      const value: unknown = _value
+      if (typeof value === 'object' && value !== null) {
+        objList = [...objList, ...getGapsWithinObject(value)]
+      }
+    })
+
+    return objList
```
Comment (hejtful) [inline src/serlo-editor/plugins/fill-in-the-gap-exercise/renderer.tsx]: I thought that using reduce would reduce (ba-dum-reducce) the amount of code here. In the end, it kind of didn't 😄 Still, here's the code, take it or leave it
```suggestion
    return Object.values(pluginState).reduce(
      (gaps: t.TypeOf<typeof GapState>[], value: unknown) => {
        if (typeof value === 'object' && value !== null) {
          gaps.concat(getGapsWithinObject(value))
        }
        return gaps
      },
      []
    )
```
(btw, renamed `obj` to `pluginState` for clarity)

File: src/serlo-editor/plugins/fill-in-the-gap-exercise/renderer.tsx [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+  )
+
+  function checkAnswers() {
+    if (mode === 'fill-in-the-gap') {
+      const newGapAnswersCorrectList = new Map<
+        GapId,
+        { isCorrect: boolean | undefined }
+      >()
+      gapStateList.forEach((gapState) => {
+        const textUserTypedIntoThisGap =
```
Comment (hejtful) [inline src/serlo-editor/plugins/fill-in-the-gap-exercise/renderer.tsx]:
```suggestion
        const userInput =
```

File: src/serlo-editor/plugins/fill-in-the-gap-exercise/renderer.tsx [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+          textUserTypedIntoGap.get(gapState.gapId)?.text ?? ''
+        const answerCorrect =
```
Comment (hejtful) [inline src/serlo-editor/plugins/fill-in-the-gap-exercise/renderer.tsx]:
```suggestion
        const isCorrect =
```

**Review by hejtful — COMMENTED** [review]
All of my comments are minor and just suggestions, you decide if they make sense for you. In general, the PR looks good to me, well done!
Didn't approve yet, because so many TODOs and commented lines make it seem like it's still a WIP PR.

File: src/serlo-editor/plugins/fill-in-the-gap-exercise/components/gap-solution-area.tsx (hunk 6 above)
Comment (LarsTheGlidingSquirrel) [reply src/serlo-editor/plugins/fill-in-the-gap-exercise/components/gap-solution-area.tsx]: This is commented out code from my experiments getting drag & drop to work. I chose to leave it in because I'm going to built upon that in the next iteration. But it definitely makes the code more messy right now.

File: src/serlo-editor/plugins/fill-in-the-gap-exercise/components/gap-solution.tsx (hunk 7 above)
Comment (LarsTheGlidingSquirrel) [reply src/serlo-editor/plugins/fill-in-the-gap-exercise/components/gap-solution.tsx]: This is commented out code from my experiments getting drag & drop to work. I chose to leave it in because I'm going to built upon that in the next iteration. But it definitely makes the code more messy right now.

File: src/serlo-editor/plugins/fill-in-the-gap-exercise/context/gap-context.ts (hunk 8 above)
Comment (LarsTheGlidingSquirrel) [reply src/serlo-editor/plugins/fill-in-the-gap-exercise/context/gap-context.ts]: I find reading `Map<GapId, ...>` is more clear than `Map<string, ...>`. Other than that the type declaration has not other use. But that is just a small thing for me, we can change it if you prefer the other.

Comment (hejtful) [reply src/serlo-editor/plugins/fill-in-the-gap-exercise/context/gap-context.ts]: Nope, all good, I just wanted to know the reasoning, thanks for explaining!

File: src/serlo-editor/plugins/text/hooks/use-slate-render-handlers.tsx (hunk 11 above)
Comment (LarsTheGlidingSquirrel) [reply src/serlo-editor/plugins/text/hooks/use-slate-render-handlers.tsx]: Changed it to `e`

File: src/serlo-editor/plugins/text/hooks/use-slate-render-handlers.tsx (hunk 12 above)
Comment (LarsTheGlidingSquirrel) [reply src/serlo-editor/plugins/text/hooks/use-slate-render-handlers.tsx]: Good idea! However, this input is only visible in the editor. The input in the no-edit view will size itself based on the users input (not based on the correct answer), so the size will not give anything away there. However, for drag and drop this might become important later because one could find the correct solution by comparing the size of the gap to the size of the draggable element.

File: src/serlo-editor/plugins/fill-in-the-gap-exercise/renderer.tsx (hunk 14 above)
Comment (LarsTheGlidingSquirrel) [reply src/serlo-editor/plugins/fill-in-the-gap-exercise/renderer.tsx]: I like it! Thank you for the suggestion.

File: src/serlo-editor/editor-ui/plugin-toolbar/text-controls/utils/math.ts (hunk 4 above)
Comment (elbotho) [reply src/serlo-editor/editor-ui/plugin-toolbar/text-controls/utils/math.ts:45]: nice, thanks for looking it up.

File: src/serlo-editor-integration/create-renderers.tsx [hunk 17]
```diff
@@ -6,27 +6,29 @@ import { GeogebraSerloStaticRenderer } from './serlo-plugin-wrappers/geogebra-se
...
-  EditorPageLayoutDocument,
-  EditorPagePartnersDocument,
-  EditorPageTeamDocument,
-  EditorScMcExerciseDocument,
-  EditorSerloTableDocument,
-  EditorSolutionDocument,
-  EditorSpoilerDocument,
-  EditorTemplateExerciseGroupDocument,
+import {
+  EditorFillInTheGapExerciseDocument,
```
Comment (elbotho) [inline src/serlo-editor-integration/create-renderers.tsx]:
```suggestion
  type EditorFillInTheGapExerciseDocument,
```

Comment (LarsTheGlidingSquirrel) [reply src/serlo-editor-integration/create-renderers.tsx]: Good catch

File: src/data/en/index.ts (hunk 1 above)
Comment (LarsTheGlidingSquirrel) [reply src/data/en/index.ts]: It is "fill in the blanks" (like in H5P) now

File: src/serlo-editor/plugins/fill-in-the-gap-exercise/renderer.tsx (hunk 15 above)
Comment (LarsTheGlidingSquirrel) [reply src/serlo-editor/plugins/fill-in-the-gap-exercise/renderer.tsx]: There will be a `solutionUserDroppedIntoThisBlank` later in addition to `textUserTypedIntoThisBlank`. I think especially because of the two modes it makes sense to name them more explicitly than `userInput`.

**Comment by elbotho** [comment]
>  Removing gap should keep text instead of deleting it.

While checking this out I realized that links disappear in the `fill in the gaps` exercise as well. the text just disappears when activating a link.

File: src/serlo-editor/plugins/text/plugins/with-blanks.ts [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+
+export const withBlanks = (editor: Editor) => {
+  const { isInline } = editor
+
+  editor.isInline = (element) => {
+    return element.type === 'blank' ? true : isInline(element)
+  }
+
+  editor.isVoid = (element) => {
+    return element.type === 'blank' ? true : isInline(element)
```
Comment (elbotho) [inline src/serlo-editor/plugins/text/plugins/with-blanks.ts]:
```suggestion
    return element.type === 'blank' ? true : isVoid(element)
```
This should solve the link problem I think

**Review by elbotho — APPROVED** [review]
I addressed the points that were important for me in my PR, so I’m fine with merging now.
Most todos are now (more or less) contained in the following files, that should be fine imho:
- `src/serlo-editor/plugins/text/hooks/use-slate-render-handlers.tsx`
- `src/serlo-editor/plugins/fill-in-the-blanks-exercise/…`
